core/src/test/java/org/apache/calcite/test/CalciteAssert.java

```diff
@@ -933,185 +925,170 @@ public AssertMetaData metaData(Function<Connection, ResultSet> function) {
     }
   }
 
-  /** Connection factory. */
-  public interface ConnectionFactory {
-    Connection createConnection() throws Exception;
-    ConnectionFactory with(String property, String value);
-    ConnectionFactory with(String schemaName, Schema schema);
-  }
+  /**
+   * Abstract implementation of connection factory whose {@code with}
+   * methods throw.
+   *
+   * <p>Avoid creating new sub-classes otherwise it would be hard to support
+   * {@code .with(property, value).with(...)} kind of chains.
+   *
+   * <p>If you want augment the connection, use {@link ConnectionPostProcessor}.
+   **/
+  public abstract static class ConnectionFactory {
+    public abstract Connection createConnection() throws SQLException;
 
-  /** Abstract implementation of
-   * {@link org.apache.calcite.test.CalciteAssert.ConnectionFactory}
-   * whose {@code with} methods throw. */
-  public abstract static class AbstractConnectionFactory
-      implements ConnectionFactory {
     public ConnectionFactory with(String property, String value) {
       throw new UnsupportedOperationException();
     }
 
-    public ConnectionFactory with(String schemaName, Schema schema) {
+    public ConnectionFactory with(ConnectionPostProcessor postProcessor) {
       throw new UnsupportedOperationException();
     }
   }
 
-  /** Connection factory that uses the same instance of connections. */
-  private static class PoolingConnectionFactory implements ConnectionFactory {
-    private final ConnectionFactory factory;
+  /** Connection post processor */
+  public interface ConnectionPostProcessor
+      extends Function<Connection, Connection> {
+  }
 
-    public PoolingConnectionFactory(final ConnectionFactory factory) {
-      this.factory = factory;
-    }
+  /** Adds {@link Schema} and sets it as default. */
+  public static class AddSchemaPostProcessor
+      implements ConnectionPostProcessor {
+    private final String name;
+    private final Schema schema;
 
-    public Connection createConnection() throws Exception {
-      return Pool.INSTANCE.cache.get(factory);
+    public AddSchemaPostProcessor(String name, Schema schema) {
+      this.name = name;
+      this.schema = schema;
     }
 
-    public ConnectionFactory with(String property, String value) {
-      throw new UnsupportedOperationException();
+    public Connection apply(Connection input) {
+      try {
+        if (schema != null) {
+          CalciteConnection con = input.unwrap(CalciteConnection.class);
+          SchemaPlus rootSchema = con.getRootSchema();
+          rootSchema.add(name, schema);
+        }
+        Compatible.INSTANCE.setSchema(input, name);
+      } catch (SQLException e) {
+        Throwables.propagate(e);
```

Comment:
Can you please `throw Throwables.propagate(e)` to make the control flow explicit?